code/game/gamemodes/cult/blood_magic.dm

```diff
@@ -735,148 +735,163 @@
+		return
+	if(H.AmountCultSlurring())
+		to_chat(user, "<span class='danger'>[H.p_their(TRUE)] blood has been tainted by an even stronger form of blood magic, it's no use to us like this!</span>")
+		return
+	if(!H.dna || (NO_BLOOD in H.dna.species.species_traits) || H.dna.species.exotic_blood != null)
+		to_chat(user, "<span class='warning'>[H] does not have any usable blood!</span>")
+		return
+	if(H.blood_volume <= BLOOD_VOLUME_SAFE)
+		to_chat(user, "<span class='warning'>[H] is missing too much blood - you cannot drain [H.p_them()] further!</span>")
+		return
+	H.blood_volume -= 100
+	uses += 50
+	user.Beam(H, icon_state = "drainbeam", time = 10)
+	playsound(get_turf(H), 'sound/misc/enter_blood.ogg', 50)
+	H.visible_message("<span class='danger'>[user] has drained some of [H]'s blood!</span>",
+					"<span class='userdanger'>[user] has drained some of your blood!</span>")
+	to_chat(user, "<span class='cultitalic'>Your blood rite gains 50 charges from draining [H]'s blood.</span>")
+	new /obj/effect/temp_visual/cult/sparks(get_turf(H))
+	return
 
-		..()
+// This should really be split into multiple procs
+/obj/item/melee/blood_magic/manipulator/afterattack(atom/target, mob/living/carbon/human/user, proximity)
+	if(user.holy_check())
+		return
+	if(!proximity)
+		return ..()
+	if(ishuman(target))
+		if(iscultist(target))
+			heal_cultist(user, target)
+			target.clean_blood()
+		else
+			steal_blood(user, target)
+		return
+
+	if(isconstruct(target))
+		heal_construct(user, target)
+		return
+
+	if(istype(target, /obj/item/blood_orb))
+		var/obj/item/blood_orb/candidate = target
+		if(candidate.blood)
+			uses += candidate.blood
+			to_chat(user, "<span class='warning'>You obtain [candidate.blood] blood from the orb of blood!</span>")
+			playsound(user, 'sound/misc/enter_blood.ogg', 50)
+			qdel(candidate)
+			return
+	blood_draw(target, user)
 
 /obj/item/melee/blood_magic/manipulator/proc/blood_draw(atom/target, mob/living/carbon/human/user)
 	var/temp = 0
-	var/turf/T = get_turf(target)
-	if(T)
-		for(var/obj/effect/decal/cleanable/blood/B in view(T, 2))
-			if(B.blood_state == BLOOD_STATE_HUMAN && (B.can_bloodcrawl_in() || istype(B, /obj/effect/decal/cleanable/blood/slime)))
-				if(B.bloodiness == 100) //Bonus for "pristine" bloodpools, also to prevent cheese with footprint spam
-					temp += 30
-				else
-					temp += max((B.bloodiness ** 2) / 800, 1)
-				new /obj/effect/temp_visual/cult/turf/open/floor(get_turf(B))
-				qdel(B)
-		for(var/obj/effect/decal/cleanable/trail_holder/TH in view(T, 2))
-			qdel(TH)
-		if(temp)
-			user.Beam(T, icon_state = "drainbeam", time = 15)
-			new /obj/effect/temp_visual/cult/sparks(get_turf(user))
-			playsound(T, 'sound/misc/enter_blood.ogg', 50)
-			temp = round(temp)
-			to_chat(user, "<span class='cultitalic'>Your blood rite has gained [temp] charge\s from blood sources around you!</span>")
-			uses += max(1, temp)
+	var/turf/T = isturf(target) ? target : get_turf(target)
```

Comment:
using get_turf on a turf should always return the same turf
```suggestion
	var/turf/T = get_turf(target)
```

Reply:
If you're sure, why not